File: src/main/java/com/lessmarkup/engine/data/ChangeTrackerImpl.java
package com.lessmarkup.engine.data;

import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.dataobjects.EntityChangeHistory;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.cache.EntityChangeType;
import com.lessmarkup.interfaces.data.ChangeListener;
import com.lessmarkup.interfaces.data.ChangeTracker;
import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.module.Implements;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.logging.Level;
import java.util.logging.Logger;

@Implements(ChangeTracker.class)
class ChangeTrackerImpl implements ChangeTracker {

    private boolean changeTrackingInitialized;
    private final DomainModelProvider domainModelProvider;
    private final Object syncObject = new Object();
    private long lastUpdateId;
    private final Queue<EntityChangeHistory> changeQueue = new LinkedList<>();
    private final List<ChangeListener> listeners = new ArrayList<>();

    @Inject
    public ChangeTrackerImpl(DomainModelProvider domainModelProvider) {
        this.domainModelProvider = domainModelProvider;
    }
    
    private void handleQueue() {
        if (listeners.isEmpty()) {
            synchronized(syncObject) {
                if (listeners.isEmpty()) {
                    changeQueue.clear();
                    return;
                }
            }
        }
        
        for (;;) {
            EntityChangeHistory change;
            synchronized(syncObject) {
                change = changeQueue.poll();
                if (change == null) {
                    break;
                }
            }
            listeners.stream().forEach(listener -> {
                listener.onChange(change.getId(), change.getUserId(), change.getEntityId(), change.getCollectionId(), EntityChangeType.of(change.getChangeType()));
            });
        }
    }
    
    public void stop() {
    }
    
    private void initializeChangeTracker() {
        if (changeTrackingInitialized) {
            return;
        }
        
        synchronized(syncObject) {
            if (changeTrackingInitialized) {
                return;
            }
            
            changeTrackingInitialized = true;
            
            try (DomainModel domainModel = domainModelProvider.create()) {
                EntityChangeHistory history = domainModel.query().from(EntityChangeHistory.class).orderByDescending(Constants.Data.ID_PROPERTY_NAME).firstOrDefault(EntityChangeHistory.class);
                if (history != null) {
                    lastUpdateId = history.getId();
                }
            } catch (Exception ex) {
                Logger.getLogger(ChangeTrackerImpl.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
    }
    
    @Override
    public void invalidate() {
        enqueueUpdates();
        handleQueue();
    }

    @Override
    public <T extends DataObject> void addChange(Class<T> type, long objectId, EntityChangeType changeType, DomainModel domainModel) {
        EntityChangeHistory record = new EntityChangeHistory();
        record.setEntityId(objectId);
        record.setChangeType(changeType.ordinal());
        record.setUserId(RequestContextHolder.getContext().getCurrentUser().getUserId());
        record.setCollectionId(DomainModelImpl.getCollectionId(type).getAsInt());
        record.setCreated(OffsetDateTime.now());
        domainModel.create(record);
    }

    @Override
    public <T extends DataObject> void addChange(Class<T> type, T dataObject, EntityChangeType changeType, DomainModel domainModel) {
        initializeChangeTracker();
        EntityChangeHistory record = new EntityChangeHistory();
        record.setEntityId(dataObject.getId());
        record.setChangeType(changeType.ordinal());
        record.setUserId(RequestContextHolder.getContext().getCurrentUser().getUserId());
        record.setCollectionId(DomainModelImpl.getCollectionId(type).getAsInt());
        record.setCreated(OffsetDateTime.now());
        domainModel.create(record);
    }

    @Override
    public void subscribe(ChangeListener listener) {
        initializeChangeTracker();
        listeners.add(listener);
    }

    @Override
    public void unsubscribe(ChangeListener listener) {
        initializeChangeTracker();
        listeners.remove(listener);
    }

    @Override
    public void enqueueUpdates() {
        try (DomainModel domainModel = domainModelProvider.create()) {
            domainModel.query().from(EntityChangeHistory.class).where("id > " + lastUpdateId).orderBy("id").toList(EntityChangeHistory.class).forEach(h -> {
                lastUpdateId = h.getId();
                synchronized(syncObject) {
                    changeQueue.add(h);
                }
            });
        } catch (Exception ex) {
            Logger.getLogger(ChangeTrackerImpl.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
}

File: src/main/scala/com/lessmarkup/engine/data/ChangeTrackerImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.data

import java.time.OffsetDateTime
import java.util.logging.{Level, Logger}

import com.google.inject.Inject
import com.lessmarkup.Constants
import com.lessmarkup.dataobjects.EntityChangeHistory
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.Implements
import com.lessmarkup.interfaces.cache.EntityChangeType
import com.lessmarkup.interfaces.data.{ChangeListener, ChangeTracker, DataObject, DomainModel, DomainModelProvider}

import scala.collection.JavaConversions._
import scala.collection.mutable

@Implements(classOf[ChangeTracker])
class ChangeTrackerImpl @Inject() (domainModelProvider: DomainModelProvider) extends ChangeTracker {

  private var changeTrackingInitialized: Boolean = false
  private final val syncObject: AnyRef = new AnyRef
  private var lastUpdateId: Long = 0L
  private val changeQueue: mutable.ListBuffer[EntityChangeHistory] = mutable.ListBuffer()
  private var listeners: List[ChangeListener] = List()

  private def handleQueue() {

    if (listeners.isEmpty) {
      syncObject synchronized {
        if (listeners.isEmpty) {
          changeQueue.clear()
          return
        }
      }
    }

    while (changeQueue.nonEmpty) {
      var change: Option[EntityChangeHistory] = None
      syncObject synchronized {
        change = changeQueue.headOption
        if (change.isDefined)
          changeQueue.remove(0)
      }
      if (change.isDefined) {
        listeners.foreach(listener => listener.onChange(
          change.get.id,
          change.get.userId,
          change.get.entityId,
          change.get.collectionId,
          EntityChangeType(change.get.changeType)))
      }
    }
  }

  def stop() {
  }

  private def initializeChangeTracker() {
    if (changeTrackingInitialized) {
      return
    }
    syncObject synchronized {
      if (changeTrackingInitialized) {
        return
      }
      changeTrackingInitialized = true
      val domainModel: DomainModel = domainModelProvider.create
      try {
        val history: Option[EntityChangeHistory] = domainModel.query
          .from(classOf[EntityChangeHistory])
          .orderByDescending(Constants.DataIdPropertyName)
          .first(classOf[EntityChangeHistory], None)
        if (history.isDefined) {
          lastUpdateId = history.get.id
        }
      }
      catch {
        case ex: Exception =>
          Logger.getLogger(classOf[ChangeTrackerImpl].getName).log(Level.SEVERE, null, ex)
      } finally {
        if (domainModel != null) domainModel.close()
      }
    }
  }

  def invalidate() {
    enqueueUpdates()
    handleQueue()
  }

  def addChange[T <: DataObject](`type`: Class[T], objectId: Long, changeType: EntityChangeType, domainModel: DomainModel) {
    val record: EntityChangeHistory = new EntityChangeHistory(
      entityId = objectId,
      changeType = changeType.value,
      userId = RequestContextHolder.getContext.getCurrentUser.getUserId,
      collectionId = MetadataStorage.getCollectionId(`type`).get,
      created = OffsetDateTime.now
    )
    domainModel.create(record)
  }

  def addChange[T <: DataObject](`type`: Class[T], dataObject: T, changeType: EntityChangeType, domainModel: DomainModel) {
    initializeChangeTracker()
    val record = new EntityChangeHistory(
      entityId = dataObject.id,
      changeType = changeType.value,
      userId = RequestContextHolder.getContext.getCurrentUser.getUserId,
      collectionId = MetadataStorage.getCollectionId(`type`).get,
      created = OffsetDateTime.now
    )
    domainModel.create(record)
  }

  def subscribe(listener: ChangeListener) {
    initializeChangeTracker()
    listeners = listener :: listeners
  }

  def unsubscribe(listener: ChangeListener) {
    initializeChangeTracker()
    listeners = listeners.filter(_ != listener)
  }

  def enqueueUpdates() {
    val domainModel: DomainModel = domainModelProvider.create
    try {
      domainModel.query
        .from(classOf[EntityChangeHistory])
        .where("id > " + lastUpdateId).orderBy("id")
        .toList(classOf[EntityChangeHistory])
        .foreach(h => {
        lastUpdateId = h.id
        syncObject synchronized {
          changeQueue.add(h)
        }
      })
    }
    catch {
      case ex: Exception =>
        Logger.getLogger(classOf[ChangeTrackerImpl].getName).log(Level.SEVERE, null, ex)
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }
}
